### PR TITLE
Happychat: Prevent Chat bubble from appearing always for all users

### PR DESCRIPTION
**Pull request by mattwondra**
Currently, all users are seeing the chat bubble appear in the left sidebar's footer once Happychat connects, whether they should or not. This is due to some regressions in Happychat selector logic and a lack of clarity in those selectors' intent.

This PR cleans up those selectors so that only users in an active Happychat conversation will see the chat bubble.

Refs: 522-gh-happychat + 440-gh-happychat

### To reproduce

- In `master` sign in as a free user or create a new free account
- Navigate to `/help/contact`
- You should see the Happychat bubble appear (this should not happen)

### To test the fix

- Reproduce steps above with `happychat/fix-chat-bubble-logic` branch, and the chat bubble should not appear

### To test against regressions

- Sign in as a paid user and start a Happychat session
- The bubble should appear
- Refresh the page — the bubble should still be there (unless Calypso reset your Redux state on refresh)
- `/end` the chat from the operator and refresh Calypso — you should not see the bubble

### Check selector logic

These selectors were modified: `hasActiveHappychatSession`, `isHappychatAvailable`, `canUserSendMessages`

Read the docblocks added above the modified selectors — do the selector name and logic seem to do what the description says it should do?

Check each usage in the code of those selectors (there's not _that_ many). Does the intention of the selector match how its being used in each instance?

### CODE REVIEW

**Comment by southp** [comment]
Not sure if it's only me, somehow I couldn't reproduce the issue :/

File: client/state/happychat/selectors.js [hunk 1]
```diff
@@ -97,16 +112,25 @@ export const getHappychatTimeline = createSelector(
...
-	[ getHappychatConnectionStatus, getHappychatChatStatus ]
+/**
+ * Returns true if the user should be able to send messages to operators based on
+ * chat status. For example new chats and ongoing chats should be able to send messages,
+ * but blocked or pending chats should not.
+ * @param {Object} state - global redux state
+ * @return {Boolean} Whether the user is able to send messages
+ */
+export const canUserSendMessages = state => (
+	isHappychatAvailable( state ) &&
```
Comment (omarjackman) [inline client/state/happychat/selectors.js]: We no longer care about the happychat connection status here?

**Comment by omarjackman** [comment]
I left one inline comment/question but I also notice that the tests for this file are looking a little light. Should we try to add a test for the new method you introduced `hasActiveHappychatSession`

**Comment by omarjackman** [comment]
Also maybe add a test for `canUserSendMessages` since it looks like its now disregarding the happychat connection status

File: client/state/happychat/selectors.js (hunk 1 above)
Comment (mattwondra) [inline client/state/happychat/selectors.js]: Connection status is implicitly included as part of the `isHappychatAvailable` selector, which makes the same check on `isHappychatClientConnected` that was here before.

We _could_ explicitly include an `isHappychatClientConnected` check here, but I don't think we should — we should trust that `isHappychatAvailable` means "Happychat is able to take chats" regardless of connection state (for example, maybe in the future we allow users to write messages before a connection has been established — Happychat could still be considered "available" even without an active connection).

**Comment by mattwondra** [comment]
@southp Hm, was Calypso pointed at an active Happychat server? (e.g. if using local, was your local Happychat up?)

Otherwise I'm not sure why it wouldn't trigger... You can trigger it on WordPress.com right now too, which should reliably reproduce. Here's a GIF of me triggering it, starting from a new Incognito window, creating a new free account, and getting the chat bubble:

calypso chat bubble bug

**Comment by mattwondra** [comment]
@omarjackman Thanks for the suggestions, I just added tests.

This PR is still ready for review and hopefully unconditional approval ;)

File: client/state/happychat/selectors.js (hunk 1 above)
Comment (omarjackman) [inline client/state/happychat/selectors.js]: > we should trust that isHappychatAvailable means "Happychat is able to take chats" regardless of connection state (for example, maybe in the future we allow users to write messages before a connection has been established — Happychat could still be considered "available" even without an active connection).

That sounds like the perfect description for an automated test case 😄

File: client/state/happychat/test/selectors.js [hunk 2]
```diff
@@ -58,6 +70,59 @@ describe( 'selectors', () => {
...
+			HAPPYCHAT_CHAT_STATUS_ASSIGNED,
+			HAPPYCHAT_CHAT_STATUS_ASSIGNING,
+			HAPPYCHAT_CHAT_STATUS_CLOSED,
+			HAPPYCHAT_CHAT_STATUS_NEW,
+		];
+
+		it( 'should return false if Happychat is unavailable', () => {
+			const state = {
+				happychat: {
+					connectionStatus: 'uninitialized',
```
Comment (omarjackman) [inline client/state/happychat/test/selectors.js]: do we have constants for the connection statuses?

File: client/state/happychat/test/selectors.js [hunk 3]
```diff
@@ -58,6 +70,59 @@ describe( 'selectors', () => {
...
+					isAvailable: false,
+					chatStatus: HAPPYCHAT_CHAT_STATUS_NEW
+				}
+			};
+			expect( canUserSendMessages( state ) ).to.be.false;
+		} );
+
+		it( "should return false if Happychat is available but the chat status doesn't allow messaging", () => {
+			messagingDisabledChatStatuses.forEach( status => {
+				const state = {
```
Comment (omarjackman) [inline client/state/happychat/test/selectors.js]: Seems tedious but I try to `deepfreeze` objects like this to make sure the selector doesn't accidentally modify state.

File: client/state/happychat/test/selectors.js [hunk 4]
```diff
@@ -58,6 +70,59 @@ describe( 'selectors', () => {
...
+					happychat: {
+						connectionStatus: 'connected',
+						isAvailable: true,
+						chatStatus: status
+					}
+				};
+				expect( canUserSendMessages( state ) ).to.be.false;
+			} );
+		} );
+
```
Comment (omarjackman) [inline client/state/happychat/test/selectors.js]: Last nitpick, I promise :trollface: 

Maybe we could add a test here with
```js
it( 'Should return true if happychat is available and client is not connected', () => {
...
} );
```

Comment (mattwondra) [inline client/state/happychat/test/selectors.js]: I'll concede that for right now this is probably worth adding in — but I think longer-term it would be better if these selectors were all in different modules so that if they depend on one another they can be mocked in tests.

That is — `canUserSendMessages` does _not_ care about connection status. It just cares about results from `isHappychatAvailable` and `getHappychatChatStatus`. So in a true unit test on the logic of `canUserSendMessages`, we should be mocking `isHappychatAvailable` and `getHappychatChatStatus` to abstract away their internal logic, and not reaching into implementation details of those dependencies by shaping the state based on their internal logic.

But we can't mock those dependencies when they're all exported from the same module, so for now it's probably fine to add the test you suggested, to cover our bases.

File: client/state/happychat/test/selectors.js (hunk 2 above)
Comment (mattwondra) [inline client/state/happychat/test/selectors.js]: We don't — I didn't want to make the test cases for this PR larger by adding those and modifying a bunch of unrelated stuff. This should be addressed though.

File: client/state/happychat/selectors.js (hunk 1 above)
Comment (mattwondra) [inline client/state/happychat/selectors.js]: I mentioned this on the other comment, but just to be clear — I'd advocate that tests for `canUserSendMessages` should mock the `isHappychatAvailable` selector. As long as the intent of the `isHappychatAvailable` selector never changes (indicating whether Happychat can take chats from this user) `canUserSendMessages` shouldn't care about its implementation details including whether there's a connection or not.
